### DISCUSSION TITLE
Image<TPixel>.TryGetSinglePixelSpan(out Span<TPixel>) doesn't hint towards when it will work. When will it work?

### DISCUSSION THREAD

**Discussion by alistaircarscadden** [SixLabors/ImageSharp#1307] [Q&A] [answered]
https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Image-1.html#SixLabors_ImageSharp_Image_1_TryGetSinglePixelSpan_System_Span__0___

Questions:

1. Does it always work? Is it unpredictable? Does it always work on Images loaded from a file?
1. What is the backup plan, constructing contiguous memory myself with all of the values from GetPixelRowSpan(Int32)?

**Comment by antonfirsov** [comment, marked as answer]
> 1. Does it always work? Is it unpredictable? Does it always work on Images loaded from a file?

Currently the limit is `int.MaxValue / 4 bytes` which is around 134 `Rgba32` megapixels. We may decide to lower this value, but it we won't go below 16 Megapixels, to not break people working with images that need to fit on screen.

> 2. What is the backup plan, constructing contiguous memory myself with all of the values from GetPixelRowSpan(Int32)?

```C#
foreach (Memory<Rgba32> memory in image.GetPixelMemoryGroup())
{
    // Take pixel data from memory.Span
}
```

PS:
These questions made me to consider exposing methods like`image.CopyPixelDataTo(Span<TPixel>)` and maybe even `image.CopyPixelDataTo(Stream)`. Is the second thing any useful?

**Reply by alistaircarscadden** [reply]
Thanks!

I think the limit should be part of the documentation of the Try... function.

Those methods could perhaps help with people trying to load an image and pass in byte[] of RGBA to OpenGL, like me.

**Reply by JimBobSquarePants** [reply]
Feel free to submit a PR adding that documentation.

**Reply by antonfirsov** [reply]
The current value is quite arbitrary one, didn't want to document it before we make a final decision.

The conclusion for the 16MP lower limit came yesterday, thanks to these discussions. There are many users like you, interoping with GPU, where having a single buffer is very helpful. My guess is that most of those use cases deal with on-screen renderable images which probably won't go over `4000 x 4000 px`.

**Reply by alistaircarscadden** [reply]
Hello @antonfirsov,
I wanted to pitch in one more idea before you set 16MP in stone. I think some current and future GPUs support 4096x4096 textures, which comes out to just over 16MP, so when you decide what to support, perhaps keep that in mind.

**Reply by antonfirsov** [reply]
Thanks, that's a good insight. I wonder if it goes above anytime soon? Maybe we should fix it for `8192 x 8192 ~ 64 MP` ?

**Reply by alistaircarscadden** [reply]
I don't know what the situation about 8192x textures looks like to be honest. I think even on cards that support 4096x textures they are not often used for much. Perhaps for 4K monitors, maybe that's what you'd need to look in to to find out if 8K will become a reality soon.

**Comment by antonfirsov** [comment]
@alistaircarscadden @ThomasMiz @Veriara and others interested in contiguous image buffers:

Despite my promise of not going below 16 Rgba32 Megapixels, we are about to radically reduce the default contiguous buffer size in our next release.

This seems to be necessary in order to optimize memory management for common use cases. We are about to provide an easy workaround to setup higher contiguous minimum for interop scenarios:
https://github.com/SixLabors/ImageSharp/issues/1596#issuecomment-886218102

**Reply by ThomasMiz** [reply]
Thanks a million for the heads up! Is there any discussion/issue/pr I can subscribe to for news on how this progresses?

**Reply by antonfirsov** [reply]
We track the work in #1596, comments, API suggestions are welcome.
